<launch>
  <arg name="open_rviz" default="true"/>
  <arg name="first_tb3"  value="tb3_0"/>
  <arg name="second_tb3" value="tb3_1"/>


  <include file="$(find turtlebot3_gazebo)/launch/multi_turtlebot3.launch">
    <arg name="first_tb3"  value="$(arg first_tb3)"/>
    <arg name="second_tb3" value="$(arg second_tb3)"/>
  </include>


  <!--launch gmapping w laser filter for each turtle bot-->
  <group ns="$(arg first_tb3)">
    <!-- Laser Filter: make inf scans appear as maxRange-->
    <!-- <node pkg="laser_filters" type="scan_to_scan_filter_chain" name="laser_filter">
      <rosparam command="load" file="$(find boundary_compute)/config/my_laser_filter.yaml"/>
      <remap from="base_scan" to="scan"/>
    </node> -->

    <node pkg="gmapping" type="slam_gmapping" name="turtlebot3_slam_gmapping" output="screen">
      <param name="base_frame" value="$(arg first_tb3)/base_footprint"/>
      <param name="odom_frame" value="$(arg first_tb3)/odom"/>
      <param name="map_frame"  value="$(arg first_tb3)/map"/>
      <rosparam command="load" file="$(find boundary_compute)/config/gmapping_params.yaml" />
      <remap from="scan" to="scan"/>
    </node>
  </group>

  <group ns="$(arg second_tb3)">
    <!-- Laser Filter: make inf scans appear as maxRange-->
    <!-- <node pkg="laser_filters" type="scan_to_scan_filter_chain" name="laser_filter">
      <rosparam command="load" file="$(find boundary_compute)/config/my_laser_filter.yaml"/>
      <remap from="base_scan" to="scan"/>
    </node> -->

    <node pkg="gmapping" type="slam_gmapping" name="turtlebot3_slam_gmapping" output="screen">
      <param name="base_frame" value="$(arg second_tb3)/base_footprint"/>
      <param name="odom_frame" value="$(arg second_tb3)/odom"/>
      <param name="map_frame"  value="$(arg second_tb3)/map"/>
      <rosparam command="load" file="$(find boundary_compute)/config/gmapping_params.yaml" />
      <remap from="scan" to="scan"/>
    </node>
  </group>


  
  <include file="$(find turtlebot3_gazebo)/launch/multi_map_merge.launch"/>
  <!-- Correct map origin-->
  <node pkg="boundary_compute" type="move_map_origin.py" name="move_map_origin" output="screen"/>

  <!--rviz  / Change config file-->
  <group if="$(arg open_rviz)"> 
    <node pkg="rviz" type="rviz" name="rviz" required="true" args="-d $(find turtlebot3_gazebo)/rviz/multi_turtlebot3_slam.rviz"/>
  </group>

</launch>